packages/wallets/src/components/FlowProvider/FlowProvider.tsx

```diff
@@ -0,0 +1,122 @@
+import React, {
+    createContext,
+    ReactElement,
+    ReactFragment,
+    ReactNode,
+    ReactPortal,
+    useContext,
+    useMemo,
+    useState,
+} from 'react';
+import { Formik, FormikErrors, FormikValues } from 'formik';
+
+export type TFlowProviderContext<T> = {
+    WalletScreen?: ReactNode;
+    currentScreenId: keyof T;
+    formValues: FormikValues;
+    isFinalScreen: boolean;
+    setFormValues: (
+        field: string,
+        value: unknown,
+        shouldValidate?: boolean | undefined
+    ) => Promise<FormikErrors<unknown> | void>;
+    switchScreen: (screenId: keyof T) => void;
+    switchNextScreen: () => void;
+};
+
+type FlowChildren = ReactElement | ReactFragment | ReactPortal;
+
+export type TWalletScreens = {
+    [id: string]: ReactNode;
+};
+
+export type TFlowProviderProps<T> = {
+    children: (context: TFlowProviderContext<T>) => FlowChildren;
+    initialValues: FormikValues;
+    screens: T;
+    screensOrder: (keyof T)[];
+};
+
+const FlowProviderContext = createContext<TFlowProviderContext<TWalletScreens> | null>(null);
+
+/**
+ * Hook to use the flow provider's context.
+ *
+ * @returns {TFlowProviderContext} The flow provider's context:
+ * - `currentScreenId`: The current screen's ID being shown
+ * - `isFinalScreen`: Indicates if the current screen is the final screen
+ * - `switchScreen`: Function which switches the current screen to another screen by their ID
+ * - `switchNextScreen`: Function which switches to the next screen by default. If the current screen is the final screen, it will not do anything.
+ * - `formValues`: The saved form values stored in Formik. By default it will contain the initial values passed in `initialValues` prop in the provider.
+ * - `setFormValues`: Function which allows persistence for a form value, which can be used to persist the form values for a previous screen or for the next screen.
+ * - `WalletScreen`: The rendered screen which is rendered by the FlowProvider.
+ */
+export const useFlow = () => {
+    const flowProviderContext = useContext(FlowProviderContext);
+
+    if (!flowProviderContext) throw new Error('useFlow must be used within a FlowProvider component.');
+
+    return flowProviderContext;
+};
+
+function FlowProvider<T extends TWalletScreens>({
+    children,
+    initialValues,
+    screens,
+    screensOrder,
+}: TFlowProviderProps<T>) {
+    const [currentScreenId, setCurrentScreenId] = useState<keyof T>(screensOrder[0]);
+    const switchScreen = (screenId: keyof T) => {
+        setCurrentScreenId(screenId);
+    };
+
+    const FlowProvider = FlowProviderContext.Provider as React.Provider<TFlowProviderContext<T> | null>;
+    const currentScreenIndex = useMemo(() => screensOrder.indexOf(currentScreenId), [currentScreenId]);
+    const isFinalScreen = currentScreenIndex >= screensOrder.length;
```

Comment:
is it ever gonna be true?
currentScreenIndex >= (screensOrder.length - 1) ? 

if its not being used, then YAGNI?

Reply:
It's not used at the moment, but I should have made it like what you suggested with the `- 1` 😅 . I can exclude out exporting `isFinalScreen` from the context, later when we need it you guys can just expose it again